npm/pkg/dataplane/dataplane_windows.go

```diff
@@ -1,33 +1,164 @@
 package dataplane
 
 import (
-	"github.com/Azure/azure-container-networking/npm"
+	"fmt"
+
+	"github.com/Azure/azure-container-networking/npm/pkg/dataplane/policies"
 	"github.com/Microsoft/hcsshim/hcn"
 	"k8s.io/klog"
 )
 
-const (
-	// Windows specific constants
-	AzureNetworkName = "azure"
-)
-
 // initializeDataPlane will help gather network and endpoint details
 func (dp *DataPlane) initializeDataPlane() error {
-	klog.Infof("Initializing dataplane for windows")
+	klog.Infof("[DataPlane] Initializing dataplane for windows")
 
-	// Get Network ID
-	network, err := hcn.GetNetworkByName(AzureNetworkName)
+	err := dp.setNetworkIDByName(AzureNetworkName)
 	if err != nil {
 		return err
 	}
 
-	dp.networkID = network.Id
+	err = dp.refreshAllPodEndpoints()
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (dp *DataPlane) shouldUpdatePod() bool {
+	return true
+}
+
+// updatePod has two responsibilities in windows
+// 1. Will call into dataplane and updates endpoint references of this pod.
+// 2. Will check for existing applicable network policies and applies it on endpoint
+func (dp *DataPlane) updatePod(pod *UpdateNPMPod) error {
+	klog.Infof("[DataPlane] updatePod called for %s/%s", pod.Namespace, pod.Name)
+	// Check if pod is part of this node
+	if pod.NodeName != dp.nodeName {
+		klog.Infof("[DataPlane] ignoring update pod as expected Node: [%s] got: [%s]", dp.nodeName, pod.NodeName)
+		return nil
+	}
+
+	podKey := getNPMPodKey(pod.Namespace, pod.Name)
+	// Check if pod is already present in cache
+	endpoint, ok := dp.endpointCache[podKey]
+	// TODO will this throw nil pointer if not ok
+	if (!ok) || (endpoint.IP != pod.PodIP) {
+		// If the existing endpoint ID has changed, it means that the Pod has been recreated
+		// this results in old endpoint to be deleted, so we can safely ignore cleaning up policies
+		// Get endpoint for this pod
+		endpoint, err := dp.getEndpointByIP(pod.PodIP)
+		if err != nil {
+			return err
+		}
+		dp.endpointCache[podKey] = endpoint
+	}
+	// TODO now check if any of the existing network policies needs to be applied
+	// Check if the removed IPSets have any network policy references ?
+	toRemovePolicies := make(map[string]struct{})
+	for _, setName := range pod.IPSetsToRemove {
+		netpolReference, err := dp.ipsetMgr.GetNetworkPolicyReferencesBySet(setName)
+		if err != nil {
+			return err
+		}
+
+		for policyName := range netpolReference {
+			toRemovePolicies[policyName] = struct{}{}
+		}
+	}
 
+	// Now check if any of these network policies are applied on this endpoint.
+	// If yes then proceed to delete the network policy
+	for policyName := range toRemovePolicies {
+		if _, ok := endpoint.NetPolReference[policyName]; ok {
+			// Delete the network policy
+			err := dp.policyMgr.RemovePolicy(policyName, []string{endpoint.ID})
```

Comment:
delete this netpol reference from endpoint, do we need a separate loop for this ?